src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -0,0 +1,375 @@
+
+  std::vector<BaseFloat *> d_features;
+  std::vector<BaseFloat *> d_ivectors;
+  for (int i = 0; i < corr_ids.size(); ++i) {
+    d_features.push_back(d_all_features_.Data() +
+                         i * input_frames_per_chunk_ *
+                             d_all_features_.Stride());
+    d_ivectors.push_back(d_all_ivectors_.Data() + i * d_all_ivectors_.Stride());
+  }
+  int features_frame_stride = d_all_features_.Stride();
+  DecodeBatch(corr_ids, d_features, features_frame_stride,
+              n_input_frames_valid_, d_ivectors, is_last_chunk, &channels);
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::DecodeBatch(
+    const std::vector<CorrelationID> &corr_ids,
+    const std::vector<BaseFloat *> &d_features, const int features_frame_stride,
+    const std::vector<int> &n_input_frames_valid,
+    const std::vector<BaseFloat *> &d_ivectors,
+    const std::vector<bool> &is_last_chunk, std::vector<int> *channels) {
+  nvtxRangePushA("DecodeBatch");
+  std::vector<int> buf;  // TODO
+  if (!channels) {
+    channels = &buf;
+    ListIChannelsInBatch(corr_ids, channels);
+  }
+  // May already be filled before feature extraction (other overload of
+  // DecodeBatch)
+  //  if (curr_batch_ichannels_.empty()) ListIChannelsInBatch(batch);
+  if (!list_channels_first_chunk_.empty())
+    cuda_decoder_->InitDecoding(list_channels_first_chunk_);
+
+  RunNnet3(*channels, d_features, features_frame_stride, n_input_frames_valid,
+           is_last_chunk, d_ivectors);
+  RunDecoder(*channels);
+
+  BuildLatticesAndRunCallbacks(corr_ids, *channels, is_last_chunk);
+  list_channels_first_chunk_.clear();
+  nvtxRangePop();
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::BuildLatticesAndRunCallbacks(
+    const std::vector<CorrelationID> &corr_ids,
+    const std::vector<int> &channels, const std::vector<bool> &is_last_chunk) {
+  list_channels_last_chunk_.clear();
+  list_corr_id_last_chunk_.clear();
+  for (int i = 0; i < is_last_chunk.size(); ++i) {
+    if (is_last_chunk[i]) {
+      list_channels_last_chunk_.push_back(channels[i]);
+      list_corr_id_last_chunk_.push_back(corr_ids[i]);
+    }
+  }
+  cuda_decoder_->PrepareForGetRawLattice(list_channels_last_chunk_, true);
+  // Storing number of callbacks not done. Used if  WaitForLatticeCallbacks()
+  // is
+  // called
+  n_lattice_callbacks_not_done_.fetch_add(list_channels_last_chunk_.size(),
+                                          std::memory_order_acquire);
+
+  // delete data used for decoding that corr_id
+  for (int32 i = 0; i < list_channels_last_chunk_.size(); ++i) {
+    uint64_t ichannel = list_channels_last_chunk_[i];
+    CorrelationID corr_id = list_corr_id_last_chunk_[i];
+    int32 ndeleted = corr_id2channel_.erase(corr_id);
+    KALDI_ASSERT(ndeleted == 1);
+    KALDI_ASSERT(thread_pool_->tryPush(
+        {&BatchedThreadedNnet3CudaOnlinePipeline::FinalizeDecodingWrapper, this,
+         ichannel, corr_id}));
```

Comment:
When recognizing a large batch of short utterances, with a low number of `--cuda-worker-threads` (8), after a while the above assert fails. Apparently this is because the task queues filled up.

It works if the assert is changed to basically `while (!tryPush()) {usleep(1000)};`, and this should exert backpressure to slow down the decode loop properly. Does this look reasonable?

Reply:
I'll update this. Thanks.